Kernel testcase framework

kishore-sn opened the pull request:
This commit contains the initial changes for kernel test case (code coverage test) framework.

seinfra commented:
Code Rule Check Result:
os/kerneltc/kernel_tc.c:55: ERROR: [IDT_M_TAB] code indent should use tabs where possible
os/kerneltc/kernel_tc.c:55: ERROR: [IDT_M_TAB] please, no spaces at the start of a line
os/kerneltc/kernel_tc.c:56: ERROR: [IDT_M_TAB] code indent should use tabs where possible
os/kerneltc/kernel_tc.c:56: ERROR: [IDT_M_TAB] please, no spaces at the start of a line
os/kerneltc/kernel_tc.c:57: ERROR: [IDT_M_TAB] code indent should use tabs where possible
os/kerneltc/kernel_tc.c:57: ERROR: [IDT_M_TAB] please, no spaces at the start of a line
os/kerneltc/kernel_tc.c:58: ERROR: [IDT_M_TAB] code indent should use tabs where possible
os/kerneltc/kernel_tc.c:58: ERROR: [IDT_M_TAB] please, no spaces at the start of a line
os/kerneltc/kernel_tc.c:59: ERROR: [IDT_M_TAB] code indent should use tabs where possible
os/kerneltc/kernel_tc.c:59: ERROR: [IDT_M_TAB] please, no spaces at the start of a line
os/kerneltc/kernel_tc.c:61: ERROR: [IDT_M_TAB] code indent should use tabs where possible
os/kerneltc/kernel_tc.c:61: ERROR: [IDT_M_TAB] please, no spaces at the start of a line
os/kerneltc/kernel_tc.c:78: ERROR: [IDT_M_TAB] code indent should use tabs where possible
os/kerneltc/kernel_tc.c:78: ERROR: [IDT_M_TAB] please, no spaces at the start of a line
os/kerneltc/kernel_tc.c:80: ERROR: [IDT_M_TAB] code indent should use tabs where possible
os/kerneltc/kernel_tc.c:80: ERROR: [IDT_M_TAB] please, no spaces at the start of a line
os/kerneltc/kernel_tc.c:83: ERROR: [IDT_M_TAB] code indent should use tabs where possible
os/kerneltc/kernel_tc.c:85: ERROR: [IDT_M_TAB] code indent should use tabs where possible
os/kerneltc/kernel_tc.c:85: ERROR: [IDT_M_TAB] please, no spaces at the start of a line
os/kerneltc/kernel_tc.c:86: ERROR: [IDT_M_TAB] code indent should use tabs where possible
os/kerneltc/kernel_tc.c:86: ERROR: [IDT_M_TAB] please, no space before tabs
os/kerneltc/kernel_tc.c:86: ERROR: [IDT_M_TAB] please, no spaces at the start of a line
os/kerneltc/kernel_tc.c:87: ERROR: [IDT_M_TAB] code indent should use tabs where possible
os/kerneltc/kernel_tc.c:87: ERROR: [IDT_M_TAB] please, no space before tabs
os/kerneltc/kernel_tc.c:88: ERROR: [IDT_M_TAB] code indent should use tabs where possible
os/kerneltc/kernel_tc.c:88: ERROR: [IDT_M_TAB] please, no space before tabs
os/kerneltc/kernel_tc.c:89: ERROR: [IDT_M_TAB] code indent should use tabs where possible
os/kerneltc/kernel_tc.c:89: ERROR: [IDT_M_TAB] please, no space before tabs
os/kerneltc/kernel_tc.c:91: ERROR: [IDT_M_TAB] code indent should use tabs where possible
os/kerneltc/kernel_tc.c:91: ERROR: [IDT_M_TAB] please, no space before tabs
os/kerneltc/kernel_tc.c:91: ERROR: [IDT_M_TAB] please, no spaces at the start of a line
os/kerneltc/kernel_tc.c:96: ERROR: [IDT_M_TAB] code indent should use tabs where possible
os/kerneltc/kernel_tc.c:96: ERROR: [IDT_M_TAB] please, no space before tabs
os/kerneltc/kernel_tc.c:96: ERROR: [IDT_M_TAB] please, no spaces at the start of a line
os/kerneltc/kernel_tc.c:97: ERROR: [IDT_M_TAB] code indent should use tabs where possible
os/kerneltc/kernel_tc.c:97: ERROR: [IDT_M_TAB] please, no space before tabs
os/kerneltc/kernel_tc.c:97: ERROR: [IDT_M_TAB] please, no spaces at the start of a line
os/kerneltc/kernel_tc.c:98: ERROR: [IDT_M_TAB] code indent should use tabs where possible
os/kerneltc/kernel_tc.c:98: ERROR: [IDT_M_TAB] please, no space before tabs
os/kerneltc/kernel_tc.c:98: ERROR: [IDT_M_TAB] please, no spaces at the start of a line
os/kerneltc/kernel_tc.c:99: ERROR: [IDT_M_TAB] code indent should use tabs where possible
os/kerneltc/kernel_tc.c:99: ERROR: [IDT_M_TAB] please, no space before tabs
os/kerneltc/kernel_tc.c:99: ERROR: [IDT_M_TAB] please, no spaces at the start of a line
os/kerneltc/kernel_tc.c:100: ERROR: [IDT_M_TAB] code indent should use tabs where possible
os/kerneltc/kernel_tc.c:100: ERROR: [IDT_M_TAB] please, no spaces at the start of a line
os/kerneltc/kernel_tc.c:103: ERROR: [IDT_M_TAB] code indent should use tabs where possible
os/kerneltc/kernel_tc.c:103: ERROR: [IDT_M_TAB] please, no spaces at the start of a line

lokeshbv commented on os/include/tinyara/kernel_tc.h:
```diff
@@ -0,0 +1,99 @@
+
+
+/****************************************************************************
+ * Platform-Dependent "Lower-Half" Audio Driver Interfaces
```
check comments here

seinfra commented:
Code Rule Check OK.

pradeepns85 commented on os/include/tinyara/kernel_tc.h:
```diff
@@ -0,0 +1,99 @@
+/****************************************************************************
+ * Pre-processor Definitions
+ ****************************************************************************/
+/* Configuration ************************************************************/
```
can you please check this alignment

kishore-sn replied:
Done

pradeepns85 commented on os/kerneltc/kernel_tc.c:
```diff
@@ -0,0 +1,120 @@
+/****************************************************************************
+ *
+ * Copyright 2016 Samsung Electronics All Rights Reserved.
```
Year needs to be changed to 2017

kishore-sn replied:
Done

pradeepns85 commented on os/kerneltc/kernel_tc.c:
```diff
@@ -0,0 +1,120 @@
+/****************************************************************************
+ * Included Files
+ ****************************************************************************/
+#include <tinyara/config.h>
```
Cross check whether you need all these header files and unwanted ones can be removed

kishore-sn replied:
Done

pradeepns85 commented on os/kerneltc/kernel_tc.c:
```diff
@@ -0,0 +1,120 @@
+	0,							/* read */
+	0,							/* write */
+	0,							/* seek */
+	kerneltc_ioctl				/* ioctl */
```
can you plz check alignment

kishore-sn replied:
Done

pradeepns85 commented on os/kerneltc/kernel_tc.c:
```diff
@@ -0,0 +1,120 @@
+ * Name: kerneltc_ioctl
+ *
+ * Description:
+ *   The standard ioctl method.  This is where ALL of the Audio work is done.
```
Can you please check the comment description part. remove audio and add relevant description.
Also please consider adding the i/p , o/p  and return details

kishore-sn replied:
Done

pradeepns85 commented on os/kerneltc/Kconfig:
```diff
@@ -0,0 +1,11 @@
+# see kconfig-language at https://www.kernel.org/doc/Documentation/kbuild/kconfig-language.txt
+#
+
+config KERNEL_TC 
```
Would you not like to give as menu and submenu options? 
selecting KERNEL_TC menu would give options for all drivers and other modules to select or unselect

kishore-sn replied:
KERNEL_TC is overall config. 
When engineer adds new module test, we can add dependent menu items.

manoharahk commented on build/configs/sidk_s5jt200/tc/defconfig:
```diff
@@ -934,3 +934,5 @@ CONFIG_STACKMONITOR_PRIORITY=100
 CONFIG_ENABLE_UPTIME=y
 # CONFIG_SYSTEM_VI is not set
+
+CONFIG_KERNEL_TC=y
```
Where is KERNEL_TC defined?

seinfra commented:
Code Rule Check OK.

kishore-sn commented:
I modified the previous commit to remove the CONFIG_KERNEL_TC flag. 
Instead we will use the already existing CONFIG_EXAMPLES_TESTCASE_KERNEL flag to enable the kernel_tc module. This is because, this module will be tightly coupled with the application side kernel testcase code. If will be required only if the kernel testcase app is enabled in the build. 

Also, all CONFIG's for subcategories under the kernel testcases shall be common for both app side and kernel side code. For example, CONFIG_TC_KERNEL_CLOCK will enable the clock module test code in the app side and corresponding support code in kernel side.

sunghan-chang commented:
@kishore-sn @pradeepns85 Should we add it on os? not on kernel?

seinfra commented:
Code Rule Check OK.

kishore-sn commented:
@sunghan-chang I added it in os because this will act as an entry point for testing all modules under os, like
os/drivers/...
os/audio/...
os/kernel/...
etc..

Please let me know if you have a different opinion.

sunghan-chang commented:
@kishore-sn There is no audio and drivers in kernel tc. If you want to add another module also, you should modify folder name. But is it enough important putting it under os? not each module?

sunghan-chang commented:
@kishore-sn Your kernel means kernel space?

kishore-sn commented:
Actually, I am talking about the kernel code coverage part. Since the ctc tools shows low code coverage for kernel part, we need to write test cases for some of the kernel modules like below:
os/drivers/...
os/kernel/...
os/fs/...
etc

In the present approach for kernel test cases, the kernel module API's are being called directly from the app side. We want to avoid this. All calls to kernel API's must go through standard interface like system calls. So, we are implementing a wrapper layer called kernel_tc. Here, we will expose a device node called /dev/kernel_tc. The test app will call ioctl's on this device node to identify a particular kernel module and a particular API to be tested.

sunghan-chang commented:
@kishore-sn I know what you do. But current testcase has kernel and it is dedicated to os/kernel and lib/libc. If you add it called kernel_tc, it can make confused. I also think that is a kernel testcase, not kernel space testcase.

sunghan-chang commented:
@kishore-sn Just testcase is enough I think. not kernel tc for name of driver.

sunghan-chang commented:
@kishore-sn And can be it called framework?

kishore-sn commented:
Well, when you say kernel testcase, it is meant to test the code residing in the kernel space right? 
When we move to the protected build, the app side testcase code cannot make direct calls to the kernel module API's. That is why we want to redesign the testcase for kernel code. As you have said, what we are proposing here is a new framework for testing kernel side code. 

We will use this framework for adding new testcases for some modules (like drivers, etc). But, we also plan to modify the already existing kernel testcases to make use of this framework. If the testcase is calling a kernel API like open, read, write, ioctl, then its fine. If it wants to call other API's in the module, then it has to be done through the new framework. 

However, I think we can rename the driver as testcase. Or how about test_fw?

sunghan-chang commented:
@kishore-sn 
```
Well, when you say kernel testcase, it is meant to test the code residing in the kernel space right?
```
1. No. As I said before, there is the "kernel" testcase under apps/examples/testcase/le_tc. So, if you make it as kernel_tc, it can make confusing.
2. There is only one testcase folder under kernel so that we don't need to call "kernel testcase". When you say where you live to your indian, you don't say "India, Bangalore", just "Bangalore". So, I think "testcase" is enough.

sunghan-chang commented on os/Directories.mk:
```diff
@@ -187,6 +187,15 @@ else
+
+# Add Kernel_tc Module
+
+ifeq ($(CONFIG_EXAMPLES_TESTCASE_KERNEL),y)
```
CONFIG_EXAMPLES_TESTCASE_KERNEL has T.C.s only for os/kernel and lib/libc. That doesn't have T.Cs for driver, audio and so on. So, if you want add whole things inside this framework, you should modify config and name.

sunghan-chang commented:
@kishore-sn Any update?

kishore-sn commented:
Hi, 
Sorry for the delay. I have made the changes as suggested. 
- Changed the module name from kerneltc to test_fw
- Added new config CONFIG_KERNEL_TEST_FW

seinfra commented:
Code Rule Check OK.

kishore-sn commented:
I removed the names starting with "kernel".

seinfra commented:
Code Rule Check OK.

sunghan-chang commented on build/configs/sidk_s5jt200/tc/defconfig:
```diff
@@ -933,3 +933,4 @@ CONFIG_STACKMONITOR_PRIORITY=100
 CONFIG_STACKMONITOR_INTERVAL=5
 CONFIG_ENABLE_UPTIME=y
 # CONFIG_SYSTEM_VI is not set
+CONFIG_KERNEL_TEST_FW=y
```
Don't modify this manually. Please update it through menuconfig.

kishore-sn replied:
Ok done

sunghan-chang commented on os/Kconfig:
```diff
@@ -464,6 +464,10 @@ menu "Audio Support"
 source audio/Kconfig
 endmenu
 
+menu "Kernel Test Framework Support"
```
1. Coudl you let me know the plan of this? I don't think this is important to be at first depth of menuconfig with current content. 
2. Without testcase example, this is not needed. Right? Then how about make it non-selectable config like ARCH_HAVE_XXX?

kishore-sn replied:
I moved it under CONFIG_EXAMPLE_TESTCASES

sunghan-chang commented on os/test_fw/test_fw.c:
```diff
@@ -0,0 +1,110 @@
+ *
+ ****************************************************************************/
+
+void test_fw_register(void)
```
Where do you call this from?

kishore-sn replied:
Added in up_initialize()

seinfra commented:
Code Rule Check OK.

sunghan-chang commented on apps/examples/testcase/Kconfig:
```diff
@@ -1,4 +1,4 @@
-#
+g
```
revert

kishore-sn replied:
Done

sunghan-chang commented on os/arch/arm/src/common/up_initialize.c:
```diff
@@ -267,5 +267,10 @@ void up_initialize(void)
 
 	up_l2ccinitialize();
+
+#ifdef CONFIG_KERNEL_TEST_FW
```
The up_xxx is an architecture specific code. But this isn't. Calling location is not correct.

kishore-sn replied:
Ok. I moved it to os/kernel/init/os_start.c

sunghan-chang commented on os/test_fw/Kconfig:
```diff
@@ -0,0 +1,10 @@
+#
+
+config KERNEL_TEST_FW
+	bool "Kernel Test Framework Support"
```
As I said before, this config should be enabled and disabled by app's testcase config automatically. Just sourcing (including) Kconfig from apps is not my opinion.

kishore-sn replied:
Ok. Added dependency to auto select KERNEL_TEST_FW based on EXAMPLES_TEST_KERNEL and EXAMPLES_TEST_DRIVER.

sunghan-chang commented on apps/examples/testcase/Kconfig:
```diff
@@ -26,6 +26,7 @@ source "$APPSDIR/examples/testcase/ta_tc/wifi_manager/utc/Kconfig"
 source "$APPSDIR/examples/testcase/ta_tc/wifi_manager/itc/Kconfig"
 source "$APPSDIR/examples/testcase/ta_tc/mqtt/utc/Kconfig"
 source "$APPSDIR/examples/testcase/ta_tc/mqtt/itc/Kconfig"
+source "$APPSDIR/../os/test_fw/Kconfig"
```
I think this is not good. Location of Kconfig should be in os, not apps because test_fw is a kernel module.

kishore-sn replied:
Ok.. I changed it to os/Kconfig and sourced it under kernel features

sunghan-chang commented:
@kishore-sn Any update?

kishore-sn commented:
We are presently working to improve the code coverage for kernel modules. 
Can we pls keep this on hold for some days?

seinfra commented:
Code Rule Check OK.

sunghan-chang commented on os/include/tinyara/fs/ioctl.h:
```diff
@@ -93,6 +93,7 @@
 #define _RTCBASE        (0x1800)	/* RTC ioctl commands */
 #define _FOTABASE       (0x1900)	/* FOTA ioctl commands */
 #define _GPIOBASE       (0x2000)	/* GPIO ioctl commands */
+#define _TESTFWIOCBASE (0xfe00)	/* KERNEL TEST FW module ioctl commands */
```
1. Please check alignments to value and comment
2. Not 0x2100? Why do you set 0xfe00?

sunghan-chang commented on os/arch/arm/src/common/up_initialize.c:
```diff
@@ -267,5 +267,6 @@ void up_initialize(void)
 	/* Initialize the L2 cache if present and selected */
 
 	up_l2ccinitialize();
+
```
Let's revert.

kishore-sn replied:
Fixed

sunghan-chang commented:
@kishore-sn Hello.
1. That is not a framework, just device driver. Isn't it? Let's remove "fw" from file and name.
2. We can move os/test_fw to os/drivers/testcase like os/drivers/audio.

sunghan-chang commented on apps/examples/testcase/le_tc/drivers/Kconfig:
```diff
@@ -6,6 +6,7 @@
 menuconfig EXAMPLES_TESTCASE_DRIVERS
 	bool "Drivers TestCase Example"
 	default n
+	select KERNEL_TEST_FW
```
Why is this needed for driver tc?

sunghan-chang commented on os/include/tinyara/test_fw.h:
```diff
@@ -0,0 +1,92 @@
+ ****************************************************************************/
+
+#include <tinyara/config.h>
+#include <tinyara/compiler.h>
```
Is this needed?

sunghan-chang commented on os/include/tinyara/test_fw.h:
```diff
@@ -0,0 +1,92 @@
+#include <tinyara/config.h>
+#include <tinyara/compiler.h>
+
+#include <tinyara/fs/ioctl.h>
```
You can put this inside CONFIG_KERNEL_TEST_FW

kishore-sn commented:
@sunghan-chang 
I have removed the use of fw from all places and changed it to driver / drv.

seinfra commented:
Code Rule Check OK.

sunghan-chang commented on os/include/tinyara/testcase_drv.h:
```diff
@@ -0,0 +1,92 @@
+ ****************************************************************************/
+
+#include <tinyara/config.h>
+#include <tinyara/compiler.h>
```
Why do you include this?

kishore-sn replied:
Fixed

sunghan-chang commented on os/drivers/testcase/testcase_drv.c:
```diff
@@ -0,0 +1,110 @@
+#include <tinyara/config.h>
+
+#include <sys/types.h>
+#include <stdbool.h>
```
check headers

kishore-sn replied:
Fixed

sunghan-chang commented on os/drivers/testcase/Makefile:
```diff
@@ -0,0 +1,69 @@
+###########################################################################
```
This file is not needed.

kishore-sn replied:
Fixed

kishore-sn commented on os/Directories.mk:
```diff
@@ -187,6 +187,7 @@ else
 KERNDEPDIRS += $(FSDIRS)
 endif
 
+
```
Fixed

kishore-sn commented on os/LibTargets.mk:
```diff
@@ -111,6 +111,12 @@ audio$(DELIM)libaudio$(LIBEXT): context
 $(LIBRARIES_DIR)$(DELIM)libaudio$(LIBEXT): audio$(DELIM)libaudio$(LIBEXT)
 	$(Q) install audio$(DELIM)libaudio$(LIBEXT) $(LIBRARIES_DIR)$(DELIM)libaudio$(LIBEXT)
 
+test_fw$(DELIM)libkerneltestfw$(LIBEXT): context
```
Fixed

seinfra commented:
Code Rule Check OK.
